docs/config.md
# 配置说明

## 配置文件路径

> 当前目录下.rshell文件夹

## 系统配置

> .rshell/cfg.yaml

```
concurrency: 6                                                       #The num of concurrency goroutine for hosts, Default: 6
tasktimeout: 300                                                     #The total timeout [second] of everyone goroutine, Default: 300
cmdseparator: ";"                                                    #The command separator, Default: ";"
promptstring: "rshell: "                                             #The system prompt string, Default: "rshell: "
outputintime: true                                                   #The output in time, Default: false
hostgroupsize: 200                                                   #The max number of ip in hostgroup, Default: 200
passcrypttype: "aes"                                                 #The password crypt type, support [aes], Default: ""
passcryptkey: "I`a!M@a#H$a%P^p&Y*c(R)y_P+t,K.eY"                     #The aes crypt key, length must be 32. Not be empty if passcrypttype = aes.
outputtype: "text"                                                   #The result output format, Default: "text"
blackcmdlist:                                                        #The dangerous black command list, Default if =cmd or ~=cmdprefix can not run.
- cmd: rm -rf /                                                      
  cmdprefix: rm -rf /                                                
mostusedcmds:                                                        #The most used command list
- pwd                                                                
- date                                                               
- whoami                                                             
updateserver:                                                        #The auto update server address list
- https://github.com/luckywinds/rshell/raw/master/releases
```

## 认证信息配置

> .rshell/auth.yaml

```
authmethods:                                                         #The authmethod list
- name: alpha-env-ttt-key-su                                         #The auth method name
  username: ttt                                                      #The username of ssh login
  privatekey: ""                                                     #First auth method option[privatekey], default: ""
  passphrase: "aes"                                                  #First auth method option[passphrase], default: ""
  password: 56efc9c43425221af1ab2ce8d6d52a6d2926ff608d56be0601d6     #Second auth method option[password], default: ""
  sudotype: "su -"                                                   #Optional：Change root command if needed, default: "su"
  sudopass: b6c2e5ba01a5ba9b1b1eae53f839c4bf13d541da17829d8408       #Optional：root password if needed, default: ""

```

> privatekey/password至少选1项即可

## 主机信息配置

> .rshell/hosts.yaml

```
hostgroups:                                 #The hostgroup list
- groupname: test-group03                   #The hostgroup name    
  authmethod: alpha-env-root-pass           #The auth method name which must in authmethods list
  sshport: 22                               #The ssh connect port, default: ""
  groups:                                   #The group list, item must in hostgroups list
  - hostgroup02
  - hostgroup03
  hosts:                                    #The host list, ip format must be ipv4
  - 192.168.31.78
  hostranges:                               #The host ip range list
  - from: 192.168.31.90
    to: 192.168.31.92
```

> groups/hosts/hostranges至少选1项即可

## 脚本配置

> 示例：example\test.yaml

```
tasks:                                      #The task list
- name: test all                            #The task name
  hostgroup: test-group01                   #The hostgroup name
  subtasks:                                 #The sub task list
  - name: test do                           #The sub task name
    mode: ssh                               #The sub task mode ssh/sftp
    cmds:                                   #The command list
    - whoami
    - date
    - echo {{.group1.abc}}
    - echo {{.group2.abc}}

  - name: test sudo                                                
    mode: ssh
    sudo: true
    cmds:
    - whoami
    - date

  - name: test upload
    mode: sftp
    ftptype: upload                         #The ftp type upload/download
    srcfile: examples/test.yaml             #The source file which will be ftp from
    desdir: .                               #The destination dir which wile be ftp to

  - name: check upload
    mode: ssh
    cmds:
    - ls -alh test.yaml

  - name: test download
    mode: sftp
    ftptype: download
    srcfile: test.yaml
    desdir: . 

```

## 变量配置

> 示例：example\values.yaml

```
group1:                                     #The var group name                          
  abc: 123                                  #The var item name
group2:
  abc: efg
```

## 执行逻辑

1. 获取系统配置
2. 获取主机配置
3. 关联认证配置

命令行交互执行模式
1. 执行相关命令

脚本编排执行模式
1. 获取变量配置，如果需要
2. 解析脚本配置，关联变量，如果需要
3. 执行相关任务命令


options/options.go
package options

import (
	"flag"
	"fmt"
	"github.com/luckywinds/rshell/pkg/checkers"
	. "github.com/luckywinds/rshell/types"
	"gopkg.in/yaml.v2"
	"html/template"
	"io/ioutil"
	"log"
	"net"
	"os"
	"path"
)

var (
	cfgFile   = path.Join(".rshell", "cfg.yaml")
	hostsFile = path.Join(".rshell", "hosts.yaml")
	authFile  = path.Join(".rshell", "auth.yaml")
	script    = flag.String("f", "", "The script yaml.")
	scriptValues   = flag.String("v", "", "The script values yaml.")
)

func init() {
	os.Mkdir(".rshell", os.ModeDir)
	initFlag()

	initCfg()
	initHostgroups()
	initAuths()
}

func initFlag() {
	cmd := os.Args[0]
	flag.Usage = func() {
		fmt.Println(`Usage:`, cmd, `[<options>]

Options:`)
		flag.PrintDefaults()
	}
	flag.Parse()
}

var cfg Cfg

func initCfg() {
	c, err := ioutil.ReadFile(cfgFile)
	if err != nil {
		log.Fatalf("Can not find cfg file[%s].", cfgFile)
	}
	err = yaml.Unmarshal(c, &cfg)
	if err != nil {
		log.Fatalf("YAML[%s] Unmarshal error: %v", cfgFile, err)
	}
}
func GetCfg() Cfg {
	if cfg.Concurrency == 0 {
		cfg.Concurrency = 6
	} else if cfg.Concurrency < 0 || cfg.Concurrency > 100 {
		log.Fatalf("Config Concurrency illegal [%d] not in (0, 100].", cfg.Concurrency)
	}
	if cfg.Tasktimeout == 0 {
		cfg.Tasktimeout = 300
	} else if cfg.Tasktimeout < 0 || cfg.Tasktimeout > 86400 {
		log.Fatalf("Config Tasktimeout illegal [%d] not in (0, 86400].", cfg.Tasktimeout)
	}
	if cfg.CmdSeparator == "" {
		cfg.CmdSeparator = ";"
	} else if len(cfg.CmdSeparator) != 1 {
		log.Fatalf("Config CmdSeparator illegal [%s] not one char.", cfg.CmdSeparator)
	}
	if cfg.PromptString == "" {
		cfg.PromptString = "rshell: "
	} else if len(cfg.PromptString) > 20 {
		log.Fatalf("Config PromptString illegal [%s] length > 20.", cfg.PromptString)
	}
	if cfg.Outputtype == "" {
		cfg.Outputtype = "text"
	} else if cfg.Outputtype != "text" {
		log.Fatalf("Config Outputtype illegal [%s] not in [text].", cfg.Outputtype)
	}
	if cfg.Hostgroupsize == 0 {
		cfg.Hostgroupsize = 200
	} else if cfg.Hostgroupsize < 0 || cfg.Hostgroupsize > 1000 {
		log.Fatalf("Config Hostgroupsize illegal [%d] not in (0, 1000].", cfg.Hostgroupsize)
	}
	if cfg.Passcrypttype != "" {
		if cfg.Passcrypttype != "aes" {
			log.Fatalf("Config Passcrypttype illegal [%s] not in [aes].", cfg.Passcrypttype)
		} else if len(cfg.Passcryptkey) != 32 {
			log.Fatalf("Config Passcryptkey illegal [%s] length != 32.", cfg.Passcryptkey)
		}
	}
	if cfg.HistoryFile == "" {
		cfg.HistoryFile = ".rshell/rshell.history"
	} else if len(cfg.HistoryFile) > 200 {
		log.Fatalf("Config HistoryFile illegal [%s] length > 200.", cfg.HistoryFile)
	}
	return cfg
}

var hostgroups Hostgroups
var hostgroupsm = make(map[string]Hostgroup)

func initHostgroups() {
	h, err := ioutil.ReadFile(hostsFile)
	if err != nil {
		log.Fatalf("Can not find hosts file[%s].", hostsFile)
	}
	err = yaml.Unmarshal(h, &hostgroups)
	if err != nil {
		log.Fatalf("YAML[%s] Unmarshal error: %v", hostsFile, err)
	}
}

func incIp(s string) string {
	ip := net.ParseIP(s)
	for j := len(ip)-1; j>=0; j-- {
		ip[j]++
		if ip[j] > 0 {
			break
		}
	}
	return ip.String()
}

func parseHosts(hg Hostgroup) Hostgroup {
	for _, h := range hg.Hosts {
		if !checkers.IsIpv4(h) {
			log.Fatalf("IP illegal [%s/%s].", hg.Groupname, h)
		}
	}
	hg.Ips = append(hg.Ips, hg.Hosts...)
	return hg
}

func parseHostrange(hg Hostgroup) Hostgroup {
	for _, hr := range hg.Hostranges {
		if !checkers.IsIpv4(hr.From) || !checkers.IsIpv4(hr.To) || hr.From == hr.To {
			log.Fatalf("IP Range illegal [%s/%s-%s].", hg.Groupname, hr.From, hr.To)
		}
		temp := []string{hr.From}
		found := false
		nip := hr.From
		count := 0
		for {
			count++
			nip = incIp(nip)
			if nip == hr.To {
				found = true
				temp = append(temp, nip)
				break
			}
			if count > cfg.Hostgroupsize && !found {
				log.Fatalf("Too Large Not Found. IP Range illegal [%s/%s-%s].", hg.Groupname, hr.From, hr.To)
			}
			temp = append(temp, nip)
		}
		if found {
			hg.Ips = append(hg.Ips, temp...)
		} else {
			log.Fatalf("IP Range illegal [%s/%s-%s].", hg.Groupname, hr.From, hr.To)
		}
	}
	return hg
}

func parseHostgroups(hgs Hostgroups) Hostgroups {
	var tmphg = Hostgroups{}
	for _, hg := range hgs.Hgs {
		hg = parseHosts(hg)
		hg = parseHostrange(hg)
		tmphg.Hgs = append(tmphg.Hgs, hg)
	}

	var hgmap = make(map[string]Hostgroup)
	for _, value := range tmphg.Hgs {
		hgmap[value.Groupname] = value
	}
	if len(tmphg.Hgs) != len(hgmap) {
		log.Fatal("There is duplicate hostgroup.")
	}

	var rethg = Hostgroups{}
	for _, hg := range tmphg.Hgs {
		for _, g := range hg.Groups {
			if hgmap[g].Groupname == "" {
				log.Fatalf("Not found. Groups illegal [%s/%s].", hg.Groupname, g)
			}
			hg.Ips = append(hg.Ips, hgmap[g].Ips...)
		}
		if checkers.IsDuplicate(hg.Ips) {
			log.Fatalf("IP Duplicate. Hostgroup illegal [%s].", hg.Groupname)
		}

		if !checkers.CheckHostgroupSize(hg, cfg.Hostgroupsize) {
			log.Fatalf("Too large. IP Range illegal [%s] > [%d].", hg.Groupname, cfg.Hostgroupsize)
		}

		if hg.Sshport < 0 {
			log.Fatalf("SSH port illegal [%d] < 0.", hg.Sshport)
		}

		if hg.Sshport == 0 {
			hg.Sshport = 22
		}
		rethg.Hgs = append(rethg.Hgs, hg)
	}

	return rethg
}

func GetHostgroups() (Hostgroups, map[string]Hostgroup) {
	if len(hostgroups.Hgs) == 0 {
		log.Fatal("The hostgroups empty.")
	}
	hostgroups = parseHostgroups(hostgroups)
	for _, value := range hostgroups.Hgs {
		hostgroupsm[value.Groupname] = value
	}

	return hostgroups, hostgroupsm
}
func GetHostgroupByname(name string) (Hostgroup, error) {
	if v, ok := hostgroupsm[name]; !ok {
		return v, fmt.Errorf("hostgroup not exist")
	} else {
		return v, nil
	}
}

var auths Auths
var authsm = make(map[string]Auth)

func initAuths() {
	a, err := ioutil.ReadFile(authFile)
	if err != nil {
		log.Fatalf("Can not find auth file[%s].", authFile)
	}
	err = yaml.Unmarshal(a, &auths)
	if err != nil {
		log.Fatalf("YAML[%s] Unmarshal error: %v", authFile, err)
	}
}
func GetAuths() (Auths, map[string]Auth) {
	if len(auths.As) == 0 {
		log.Fatal("The auths empty.")
	}
	for _, value := range auths.As {
		authsm[value.Name] = value
	}
	if len(auths.As) != len(authsm) {
		log.Fatal("There is duplicate auth.")
	}
	return auths, authsm
}
func GetAuthByname(name string) (Auth, error) {
	if v, ok := authsm[name]; !ok {
		return v, fmt.Errorf("auth not exist")
	} else {
		return v, nil
	}
}

var tasks Tasks

func initTasks(scriptFile string) {
	p, err := ioutil.ReadFile(scriptFile)
	if err != nil {
		log.Fatalf("Can not find script file[%s].", scriptFile)
	}

	err = yaml.Unmarshal(p, &tasks)
	if err != nil {
		log.Fatalf("YAML[%s] Unmarshal error: %v", scriptFile, err)
	}
}

var values map[string]interface{}

func initValues(scriptValues string) {
	if scriptValues == "" {
		return
	}
	p, err := ioutil.ReadFile(scriptValues)
	if err != nil {
		log.Fatalf("Can not find script file[%s].", scriptValues)
	}

	err = yaml.Unmarshal(p, &values)
	if err != nil {
		log.Fatalf("YAML[%s] Unmarshal error: %v", scriptValues, err)
	}
}

var tempScript = ".rshell/rshell.tmp"

func templateScript(script string) {
	t, err := template.ParseFiles(script)
	if err != nil {
		log.Fatalf("Parser template script file [%s] failed.", script)
	}

	f, err := os.Create(tempScript)
	if err != nil {
		log.Fatal("Create temp script file failed.")
	}
	defer f.Close()

	if err := t.Execute(f, values); err != nil {
		log.Fatal("Parser template script file task failed.")
	}
}

func GetTasks() (Tasks, bool) {
	if *script != "" {
		initValues(*scriptValues)
		if len(values) != 0 {
			templateScript(*script)
			initTasks(tempScript)
		} else {
			initTasks(*script)
		}
		if len(tasks.Ts) == 0 {
			log.Fatal("The tasks empty.")
		}

		return tasks, true
	} else {
		return tasks, false
	}
}
